torchtune/datasets/multimodal/_the_cauldron.py

```diff
@@ -0,0 +1,165 @@
+    `The Cauldron dataset<https://huggingface.co/datasets/HuggingFaceM4/the_cauldron>_`.
+
+    Image placeholders are prepended to the text in the ``Message`` content. Images in the
+    dataset are expected to be a list of a single PIL image, so they are simply passed through
+    with an optional column remapping if ``column_map`` is specified.
+
+    Args:
+        train_on_input (bool): Whether the model is trained on the user prompt or not.
+            Default is True.
+        column_map (Optional[Dict[str, str]]): a mapping to change the expected "images"
+            and "texts" column names to the actual column names in the dataset. Default is None,
+            keeping the default column names.
+        new_system_prompt (Optional[str]): if specified, prepend a system message. This can
+            serve as instructions to guide the model response. Default is None.
+
+    Raises:
+        ValueError: If ``column_map`` is provided and ``images`` not in ``column_map``, or
+            ``texts`` not in ``column_map``.
+    """
+
+    def __init__(
+        self,
+        train_on_input: bool = True,
+        column_map: Optional[Dict[str, str]] = None,
+        new_system_prompt: Optional[str] = None,
+    ):
+        self.train_on_input = train_on_input
+        self.new_system_prompt = new_system_prompt
+        if column_map is not None:
+            if "images" not in column_map:
+                raise ValueError(
+                    "column_map must contain 'images' as a key if specified"
+                )
+            if "texts" not in column_map:
+                raise ValueError(
+                    "column_map must contain 'texts' as a key if specified"
+                )
+            self._column_map = column_map
+        else:
+            self._column_map = {"images": "images", "texts": "texts"}
+
+    def __call__(self, sample: Mapping[str, Any]) -> Mapping[str, Any]:
+        messages = []
+        for message in sample[self._column_map["texts"]]:
+            messages.append(
+                Message(
+                    role="user",
+                    content=[
+                        {"type": "image"},
+                        {"type": "text", "content": message["user"]},
+                    ],
+                    masked=not self.train_on_input,
+                )
+            )
+            messages.append(
+                Message(
+                    role="assistant",
+                    content=[{"type": "text", "content": message["assistant"]}],
+                )
+            )
+
+        if self.new_system_prompt is not None:
+            messages = [
+                Message(
+                    role="system", content=self.new_system_prompt, masked=True, eot=True
+                )
+            ] + messages
+
+        return {"messages": messages, "images": sample[self._column_map["images"]]}
+
+
+def the_cauldron_dataset(
+    model_transform: Transform,
```

Comment:
Maybe I'm being dense but looking at this I have no idea how I should define the model_transform to construct one of these